Adding more tests to system transaction, demonstrating #1571

fredericDelaporte opened the pull request:
This PR add some more tests to system transaction tests. It also has a test demonstrating #1571.

fredericDelaporte commented on test/Npgsql.Tests/SystemTransactionTests.cs:
```diff
@@ -44,14 +44,16 @@ public void ExplicitEnlist()
                     conn.EnlistTransaction(Transaction.Current);
-                    Assert.That(conn.ExecuteNonQuery(@"INSERT INTO data (name) VALUES ('test')"), Is.EqualTo(1));
+                    Assert.That(conn.ExecuteNonQuery(@"INSERT INTO data (name) VALUES ('test')"), Is.EqualTo(1),
+                        "Unexpected insert rowcount");
```
I have added messages in a bunch of cases where the Assert failure could be ambiguous with another assert in the same test.

fredericDelaporte commented on test/Npgsql.Tests/SystemTransactionTests.cs:
```diff
@@ -83,31 +87,56 @@ public void EnlistOff()
+            // Scope disposed and not completed => rollback, but no enlistment, so changes should still be there.
+            using (var conn3 = OpenConnection(ConnectionStringEnlistOff))
+            {
+                Assert.That(conn3.ExecuteScalar("SELECT COUNT(*) FROM data"), Is.EqualTo(1), "Insert unexpectedly rollback-ed");
```
For actually testing `Enlist=false`, checking a rollback does not prevent data to be changed seems to me the best way.

fredericDelaporte replied:
If the `Enlist=false` does not work and enlist connections, both would be in same transaction and thus should see the same data. Conn2 would then read the uncommitted data from conn1. So it would not ascertain they are not enlisted.

fredericDelaporte commented on test/Npgsql.Tests/SystemTransactionTests.cs:
```diff
@@ -83,31 +87,56 @@ public void EnlistOff()
+                // No commit
+            }
+
+            AssertNumberOfRows(0);
```
Like `Enlist=false` but the other way round, checking the rollback case with `Enlist=true` seems to me the best way to ascertain the enlistment.

fredericDelaporte commented on test/Npgsql.Tests/SystemTransactionTests.cs:
```diff
@@ -130,21 +162,106 @@ public void TwoConnections()
         }
 
         [Test]
+        public void TwoConsecutiveConnections()
```
This usage pattern is common with NHibernate, when a single transaction scope span many NHibernate session consecutively used. If in such case there is no transaction promotion to distributed, this is better.

fredericDelaporte commented on test/Npgsql.Tests/SystemTransactionTests.cs:
```diff
@@ -130,21 +162,106 @@ public void TwoConnections()
+                    Assert.That(conn1.ExecuteNonQuery(@"INSERT INTO data (name) VALUES ('test1')"), Is.EqualTo(1),
+                        "Unexpected first insert rowcount");
+
+                    ForceEscalationToDistributedTx.Escalate(true);
```
This will cause a rollback from the prepare phase of the enlisted test ressource.

roji replied:
OK. To me this test seems very similar to TwoConnectionsRollback - except in that the enlistment triggers the rollback (via `ForceRollback()`) and not the user (by not calling `TransactionScope.Complete()`). I'm not sure if there's supposed to be a different, but in any case it makes sense having this scenario as well.

fredericDelaporte commented on test/Npgsql.Tests/SystemTransactionTests.cs:
```diff
@@ -130,21 +162,106 @@ public void TwoConnections()
+        }
+
+        [Test, Explicit("Failing test and 100 iteration loop.")]
+        public void DistributedRace()
```
Here is the test demonstrating the race condition. I was not able of demonstrating it with the simpler `TwoConnections` test. It looks like enlisting with a durable ressource is required for triggering it.

roji replied:
Huh. I'll try to find the time in the next few days to play around with this myself. If there really is a behavioral difference between volatile/durable on this I will have learned something new.

fredericDelaporte commented on test/Npgsql.Tests/SystemTransactionTests.cs:
```diff
@@ -154,21 +271,51 @@ public void TwoConnectionsWithFailure()
         }
 
         [Test]
+        public void ConnectionReuseAfterDistributedTransaction()
```
This test is for checking if supporting a null enlistment could be required for dodging some race condition. I should probably add a variation with a durable ressource enlisted, and some loops.

roji replied:
OK. In its current form this test doesn't seem very useful.

fredericDelaporte commented on test/Npgsql.Tests/SystemTransactionTests.cs:
```diff
@@ -269,7 +417,7 @@ public void ReuseConnection()
         [Test]
         public void ReuseConnectionRollback()
```
This test was looking like a copy paste of the previous one, renamed but not changed. Changes done now.

roji replied:
Oops... thanks for catching that and fixing!

fredericDelaporte commented on test/Npgsql.Tests/SystemTransactionTests.cs:
```diff
@@ -341,7 +489,7 @@ public void SchemaConnectionShouldntEnlist()
-            => Assert.That(GetNumberOfPreparedTransactions(), Is.EqualTo(0));
+            => Assert.That(GetNumberOfPreparedTransactions(), Is.EqualTo(0), "Prepared transactions found");
 
         int GetNumberOfPreparedTransactions()
```
Locally I have tried using that for adding a `AssertHasPreparedTransactions` in most distributed tests, but that assert was failing everywhere, finding nothing prepared.

roji replied:
Do you mean that you were seeing 0 transactions in pg_prepared_xacts although you were expecting one to be there? Do you consider this as evidence of the race condition you were mentioning before (or another race condition)?

fredericDelaporte replied:
> Do you mean that you were seeing 0 transactions in pg_prepared_xacts although you were expecting one to be there?

Yes.

> Do you consider this as evidence of the race condition you were mentioning before (or another race condition)?

No.

But maybe as an evidence that there is more differences between durable and volatile ressources. Enlisting many volatile ressources does not mandate promotion to distributed, provided there is no more than one durable ressource in the transaction.

> [As stated previously, a resource manager would make a volatile enlistment if it manages an in-memory, volatile resource. One of the benefits of using EnlistVolatile is that it does not force an unnecessary escalation of the transaction.](https://msdn.microsoft.com/en-us/library/ee818750.aspx)

> [As long as the System.Transactions infrastructure handles volatile resources and at most one durable resource that supports single-phase notifications, the transaction remains in the ownership of the System.Transactions infrastructure.](https://msdn.microsoft.com/en-us/library/ee818742.aspx)

The durable ressource used in the race test does not support single-phase, and with this one I see pg_prepared count raising. But for all the other tests, count stay 0.

roji replied:
OK. I now understand better how durable vs. volatile could induce this kind of behavior change (and possibly create race conditions)... The MSDTC may introduce some asynchronicity which doesn't exist when using only System.Transactions (i.e. volatile-only). The question then would be how other provider implementations deal with this kind of thing, if at all.

fredericDelaporte replied:
Yes, on NHibernate side, all our concurrency issues occurs only when the transaction is truly distributed (involving delegation to the external MsDtc process).

With NHibernate we have no "transaction commit delayed after scope disposal" issue with Sql-Server, be it through `SqlConnection` or `OdbcConnection`. But we have that with Oracle (managed driver, the others are not yet tested) and SQLite, in addition to PotsgreSql. So I think Microsoft knows what should be done but has not documented it properly, causing some other vendors to have those kind of issues. My best hope is still using a durable ressource manager instead of a volatile one. But as I have already written, I have no evidence that it is either enough or necessary.

#1580 is related to those issues: with sql-server in some circumstances re-using a connection after a transaction fails because it has not yet un-enlist from the transaction. Usually it does un-enlist by itself, but in distributed cases, sometimes it does not (or does it too late). There, enlisting with `null` allows to un-enlist from the already completed transaction and allows using the connection without a bad hack like `Thread.Sleep`.

fredericDelaporte replied:
With commit 90f01ecc341fe11ab19464e4ad7cd795ae28249f, I have tried putting back that `AssertHasPreparedTransactions` in place for cases enlisting the test durable ressource, but it fails there too. Although I have seen that count raising in some circumstances when debugging...

fredericDelaporte replied:
Well I guess I misunderstand it. I think this counter can go-up only when the prepare phase has been executed, but is not yet committed or rollback. So we cannot see that count in most circumstances. I do not think it is a good indicator for knowing if a distributed transaction is ongoing. Better check `Transaction.Current.TransactionInformation.DistributedIdentifier != Guid.Empty`.

fredericDelaporte commented on test/Npgsql.Tests/SystemTransactionTests.cs:
```diff
@@ -430,9 +623,9 @@ public void OneTimeTearDown()
-            public void Rollback(SinglePhaseEnlistment singlePhaseEnlistment) {}
+            public void Initialize() { }
+            public void SinglePhaseCommit(SinglePhaseEnlistment singlePhaseEnlistment) { }
+            public void Rollback(SinglePhaseEnlistment singlePhaseEnlistment) { }
```
Hum, looks like I have issued a ctrl+k ctrl+d.

roji replied:
Yeah, please remove these spaces

roji commented on test/Npgsql.Tests/SystemTransactionTests.cs:
```diff
@@ -130,21 +162,106 @@ public void TwoConnections()
+        public void DistributedRace()
+        {
+            var i = 0;
+            while (i < 100)
```
Nitpick: maybe a for loop?

fredericDelaporte replied:
So used to `foreach` that I do not remember there is a `for` one.

roji commented on test/Npgsql.Tests/SystemTransactionTests.cs:
```diff
@@ -296,7 +444,7 @@ public void ReuseConnectionWithEscalation()
                     conn1.Open();
                     var processId = conn1.ProcessID;
-                    using (var conn2 = new NpgsqlConnection(ConnectionStringEnlistOn)) {}
+                    using (new NpgsqlConnection(ConnectionStringEnlistOn)) { }
```
Nit: remove space between curlies

fredericDelaporte commented on test/Npgsql.Tests/SystemTransactionTests.cs:
```diff
@@ -154,20 +252,53 @@ public void TwoConnectionsWithFailure()
         }
 
+        [Test, Explicit("Failing test and 100 iteration loop.")]
+        public void ConnectionReuseRaceAfterDistributedTransaction()
```
With the test durable ressource enlisted instead of a second connection, now it fails with the error "connection already executing".

fredericDelaporte commented:
Rebased, slightly squashed, and added some more test for comparing volatile vs durable, and checking distributed vs non-distributed.

The race tests now demonstrates that even the durable resource has its commit phase happening after the scope disposal... In such a situation, the only way I can think of for avoiding a race condition when re-using the connection after the scope, it to introduced some internal locking mechanism inside it. A bit like what I have done for NHibernate session [here](https://github.com/nhibernate/nhibernate-core/pull/627/commits/ec4c68abb3574ffc57c6cbf53f8ffe1c5eefd441#diff-5462baa783fbe15a4278463a8692c7caR79). Any session usage will be blocked if required by calling the `WaitOne` added in that `IEnlistmentNotification` class.

But that will not avoid race conditions when using another connection to get at the data supposedly already committed.

I do not think I will add more to this PR now.
